### ISSUE TITLE
Volume Trajectory Build-out 

### ISSUE THREAD

**Issue by jdhoffa** [2DegreesInvesting/r2dii.analysis#132]
We currently calculate `production` * `loan_size_*`.
As a bank, we would like to calculate the build-out rate of a company, by year:
![image](https://user-images.githubusercontent.com/8741309/87449583-a7c00b00-c5fd-11ea-8f24-0fa6dbf8a2cc.png)
and use this value to multiply by `loan_size_*`.

Caveat, if a company has 0 production for t0, this will drive results to Inf. Must filter these cases out, and throw a warning if this is the case.

**Comment by jdhoffa** [comment]
@2diiKlaus

**Comment by maurolepore** [comment]
> Caveat, if a company has 0 production for t0, this will drive results to Inf. Must filter these cases out, and throw a warning if this is the case.

I wonder if there is some transformation we could use. Maybe logarithms can help, or maybe we transform 0 into a tiny number?

**Comment by jdhoffa** [comment]
So log(0) is undefined, and even if we transform 0 into a tiny number, the result will approach infinity, and will vastly skew the results. 

This reflects the more intuitive explanation: 
- the concept of "build-out" rate, is only really well-defined if the company already exists, has an existing stock of product
- any new company will have a massive build-out rate, that isn't even really useful/ relevant to measure

My preference would be to highlight this intuitive explanation (either in the function definition, or in the warning message), so that this short-coming of this approach is highlighted from the onset... Maybe let's discuss further?

@2diiKlaus I think this speaks to a broader issue actually: when do we allow the banks to decide which metrics will and won't be included, and when do we have final say (or is there a standard acceptance criteria for adding new metrics, such that it is approved conceptually by enough people, beyond a one-on-one email thread?)

**Comment by maurolepore** [comment]
Well, I'm glad you Jackson made something useful from my ignorance :)

**Comment by 2diiKlaus** [comment]
@jdhoffa yes it totally speaks to that issue and while I agree that this metric has a bunch of caveats or let's say a single caveat that can create non-sense results, it makes sense for large corporate only portfolios. 

I actually tend to not include it and try to explain it more to Imene, which I did extensively in a long e-mail chain last week or 2 weeks ago. 

Happy to discuss this - let's add it to the bank committee agenda

**Comment by jdhoffa** [comment]
@2diiKlaus it is unclear to me if build-out should be portfolio weighted or not (see Option 1 or Option 2 at the end of this reprex): 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

# a simple portfolio with only two loans to two companies (sector etc doesnt matter)
data <- tibble::tribble(
  ~name_company, ~loan_size_outstanding, ~year, ~production,
            "a",                   900L, 2020L,          1L,
            "a",                   900L, 2021L,          2L,
            "b",                   100L, 2020L,        100L,
            "b",                   100L, 2021L,        200L
  )

data
#> # A tibble: 4 x 4
#>   name_company loan_size_outstanding  year production
#>   <chr>                        <int> <int>      <int>
#> 1 a                              900  2020          1
#> 2 a                              900  2021          2
#> 3 b                              100  2020        100
#> 4 b                              100  2021        200

company_level_build_out <- data %>% 
  group_by(name_company) %>% 
  arrange(name_company, year) %>% 
  mutate(build_out = (production - first(production))/first(production)) %>% 
  ungroup()

company_level_build_out
#> # A tibble: 4 x 5
#>   name_company loan_size_outstanding  year production build_out
#>   <chr>                        <int> <int>      <int>     <dbl>
#> 1 a                              900  2020          1         0
#> 2 a                              900  2021          2         1
#> 3 b                              100  2020        100         0
#> 4 b                              100  2021        200         1

weighted_company_level_build_out <- company_level_build_out %>% 
  group_by(year) %>% 
  mutate(port_weight = loan_size_outstanding/sum(loan_size_outstanding)) %>% 
  mutate(weighted_build_out = port_weight * build_out) %>% 
  select(name_company, year, weighted_build_out)

weighted_company_level_build_out
#> # A tibble: 4 x 3
#> # Groups:   year [2]
#>   name_company  year weighted_build_out
#>   <chr>        <int>              <dbl>
#> 1 a             2020                0  
#> 2 a             2021                0.9
#> 3 b             2020                0  
#> 4 b             2021                0.1
  
# Option 1 unweighted build out at portfolio-level
company_level_build_out %>% 
  group_by(year) %>% 
  summarize(build_out = mean(build_out))
#> `summarise()` ungrouping output (override with `.groups` argument)
#> # A tibble: 2 x 2
#>    year build_out
#>   <int>     <dbl>
#> 1  2020         0
#> 2  2021         1

# Option 2 portfolio-weighted build out at portfolio-level
weighted_company_level_build_out %>% 
  group_by(year) %>% 
  summarize(weighted_build_out = mean(weighted_build_out))
#> `summarise()` ungrouping output (override with `.groups` argument)
#> # A tibble: 2 x 2
#>    year weighted_build_out
#>   <int>              <dbl>
#> 1  2020                0  
#> 2  2021                0.5
```

<sup>Created on 2020-07-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by 2diiKlaus** [comment]
@jdhoffa it should be portfolio weighted, if someone in the future does want non-portfolio weighted build-out trends he/she can use the absolute volume trends (client perspective) and calculate it themselves...

`company_level_build_out <- data %>% 
  group_by(name_company) %>% 
  arrange(name_company, year) %>% 
  **mutate(build_out = (production - first(production))/first(production)) %>%** 
  ungroup()
`

this needs to be done differently depending on the technology type (i.e. how the scenario is applied). For green technologies the denominator is the sum(first(production)) across all technologies or first(sector_production) if that exists

**Comment by jdhoffa** [comment]
Great! Thanks for the quick response @2diiKlaus

**Comment by jdhoffa** [comment]
@2diiKlaus how should we handle scenario trajectories?

**Comment by jdhoffa** [comment]
@maurolepore I will need to take care of #135 before continuing with this issue. I will open a PR momentarily, and would appreciate if you could review and merge whenever you have time.

**Comment by jdhoffa** [comment]
Closed by #141
